File: Training/Advanced_Cluster_Usage/Exercise3/Python/mp_pi.py
#!/usr/bin/env python3
#+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
# Program: mp_pi.py
#          Parallel Monte-Carlo algorithm for calculating PI using multiprocessing
#          Translated from omp_pi.c
#
# Usage:   python mp_pi.py <number_of_samples> <number_of_processes>
#+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
import sys
import math
import time
from multiprocessing import Pool

def lcg_rand(seed):
    a = 69069
    c = 1
    m = 2147483647
    seed = (a * seed + c) % m
    return seed, float(seed) / m

def calc_pi_chunk(args):
    samples, seed_base, process_id = args
    count = 0
    seed = seed_base + process_id * 1999
    
    for i in range(samples):
        seed, rx = lcg_rand(seed)
        seed, ry = lcg_rand(seed)
        x = rx
        y = ry
        z = x * x + y * y
        
        if z <= 1.0:
            count += 1
    
    return count

def main():
    if len(sys.argv) != 3:
        print("Usage: python mp_pi.py <number_of_samples> <number_of_processes>")
        sys.exit(1)
    
    total_samples = int(sys.argv[1])
    nprocesses = int(sys.argv[2])
    
    samples_per_process = total_samples // nprocesses
    remainder = total_samples % nprocesses
    tasks = []
    base_seed = 1202107158
    
    for i in range(nprocesses):
        chunk_size = samples_per_process + (1 if i < remainder else 0)
        tasks.append((chunk_size, base_seed, i))
    
    t0 = time.perf_counter()
    with Pool(processes=nprocesses) as pool:
        results = pool.map(calc_pi_chunk, tasks)
    
    total_count = sum(results)
    tf = time.perf_counter() - t0
    
    pi_estimate = 4.0 * total_count / total_samples
    
    print(f"Number of processes: {nprocesses:2d}")
    print(f"Exact value of PI: {math.pi:7.5f}")
    print(f"Estimate of PI:    {pi_estimate:7.5f}")
    print(f"Time: {tf:7.2f} sec.")

if __name__ == "__main__":
    main()

File: Training/Advanced_Cluster_Usage/Exercise3/R/mp_pi.R
#!/usr/bin/env Rscript

# Program: mp_pi.R
# Parallel Monte-Carlo PI calculation using R's parallel package

# Function to generate random numbers using LCG
lcg_rand <- function(seed) {
  a <- 69069
  c <- 1
  m <- 2147483647
  seed <- (a * seed + c) %% m
  return(list(seed = seed, value = seed / m))
}

# Function to calculate PI for a chunk of samples
calc_pi_chunk <- function(args) {
  samples <- args$samples
  seed_base <- args$seed_base
  pid <- args$pid
  
  count <- 0
  seed <- seed_base + pid * 1999
  
  for (i in 1:samples) {
    result <- lcg_rand(seed)
    seed <- result$seed
    x <- result$value
    
    result <- lcg_rand(seed)
    seed <- result$seed
    y <- result$value
    
    z <- x * x + y * y
    if (z <= 1) count <- count + 1
  }
  
  return(count)
}

# Main function
main <- function() {
  # Parse command-line arguments
  args <- commandArgs(trailingOnly = TRUE)
  if (length(args) != 2) {
    cat("Usage: Rscript mp_pi.R <number_of_samples> <number_of_processes>\n")
    quit(status = 1)
  }
  
  total_samples <- as.integer(args[1])
  nprocesses <- as.integer(args[2])
  
  # Set up parallel cluster
  library(parallel)
  cl <- makeCluster(nprocesses)
  
  # Split samples across processes
  samples_per_proc <- total_samples %/% nprocesses
  remainder <- total_samples %% nprocesses
  tasks <- list()
  base_seed <- 1202107158
  
  for (i in 1:nprocesses) {
    chunk_size <- samples_per_proc + (if (i <= remainder) 1 else 0)
    tasks[[i]] <- list(samples = chunk_size, seed_base = base_seed, pid = i - 1)
  }
  
  # Timing start
  t0 <- Sys.time()
  
  # Export functions to cluster
  clusterExport(cl, c("lcg_rand", "calc_pi_chunk"))
  
  # Run in parallel
  results <- parLapply(cl, tasks, calc_pi_chunk)
  
  # Stop cluster
  stopCluster(cl)
  
  # Combine results
  total_count <- sum(unlist(results))
  tf <- as.numeric(Sys.time() - t0, units = "secs")
  
  # Estimate PI
  pi_estimate <- 4.0 * total_count / total_samples
  
  # Output results
  cat(sprintf("Number of processes: %2d\n", nprocesses))
  cat(sprintf("Exact value of PI: %.5f\n", pi))
  cat(sprintf("Estimate of PI:    %.5f\n", pi_estimate))
  cat(sprintf("Time: %.2f sec.\n", tf))
}

# Execute main
main()
